### PR TITLE
Fix meta parent class detection

### PR DESCRIPTION
**Pull request by trasher**
Fix for #22581

Since GLPI v11 autoload rework; `is_a()` behaves differently with lower case class names: `var_dump(is_a('software', CommonDBTM::class, true))` was returning `true` in v10, and it now returns `false`.

I've added a `ucfirst()` call that would sovle issue on specific `Software` class, but that may not be enough (for example, `consumables_types` should translate to `Consumable` [missing ending "s"]); I guess that's why there were already some entries in `$key_to_itemtypes`.

I've also added a test to find `*_types` configuration entries that resolves to an empty array.

### CODE REVIEW

File: src/Glpi/Search/SearchEngine.php [hunk 1]
```diff
@@ -217,13 +217,15 @@ private static function getMetaParentItemtypesForTypesConfig(string $config_key)
             'rackable_types'       => [Enclosure::class, Rack::class],
             'socket_types'         => [Socket::class],
             'ticket_types'         => [Change::class, Problem::class, Ticket::class],
+            'consumables_types'    => [\Consumable::class],
+            'process_types'        => [\Item_Process::class],
```
Comment (trasher) [inline src/Glpi/Search/SearchEngine.php:222]: I'm not sure those should be added, but they're examples that do not work with the simple `ucfirst()` hack. Note they were not working on v10 neither.

File: src/Glpi/Search/SearchEngine.php [hunk 2]
```diff
@@ -217,13 +217,15 @@ private static function getMetaParentItemtypesForTypesConfig(string $config_key)
...
         ];
 
         if (array_key_exists($config_key, $key_to_itemtypes)) {
             return $key_to_itemtypes[$config_key];
         }
 
-        $itemclass = $matches[1];
+        $itemclass = ucfirst($matches[1]);
```
Comment (trasher) [inline src/Glpi/Search/SearchEngine.php]: This "fix" 14 cases:
'state_types', 'document_types', 'reservation_types', 'location_types', 'link_types', 'contract_types', 'printer_types', 'certificate_types', 'cluster_types', 'software_types', 'domain_types', 'agent_types', 'line_types', 'plug_types'

File: tests/functional/Glpi/Search/SearchEngineTest.php [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+use Glpi\Tests\DbTestCase;
+
+class SearchEngineTest extends DbTestCase
+{
+    public function testGetMetaParentItemtypesForTypesConfig(): void
+    {
+        global $CFG_GLPI;
+
+        $search_engine = new SearchEngine();
+        $exclusions = [
```
Comment (trasher) [inline tests/functional/Glpi/Search/SearchEngineTest.php:47]: Maybe some of those exclusions should be handled... Honestly, I do not know.

File: src/Glpi/Search/SearchEngine.php (hunk 2 above)
Comment (cedric-anne) [reply src/Glpi/Search/SearchEngine.php]: `DbUtil::getClassForItemtype()` will return the class string with the correct case if it exists, or null if it does not correspond to a valid class name.

```suggestion
        $itemclass = (new \DbUtils())->getClassForItemtype($matches[1]);
```

Comment (trasher) [reply src/Glpi/Search/SearchEngine.php]: This fix 18 cases:
'state_types', 'document_types', 'reservation_types', 'location_types', 'link_types', 'networkport_types', 'contract_types', 'printer_types', 'certificate_types', 'cluster_types', 'operatingsystem_types', 'software_types', 'domain_types', 'databaseinstance_types', 'agent_types', 'line_types', 'itemantivirus_types', 'plug_types'

Those was not handled with simple ucfirst:
'networkport_types', 'operatingsystem_types', 'databaseinstance_types', 'itemantivirus_types'

File: src/Glpi/Search/SearchEngine.php [hunk 4]
```diff
@@ -217,13 +217,16 @@ private static function getMetaParentItemtypesForTypesConfig(string $config_key)
             'rackable_types'       => [Enclosure::class, Rack::class],
             'socket_types'         => [Socket::class],
             'ticket_types'         => [Change::class, Problem::class, Ticket::class],
+            'itil_types'           => [Change::class, Problem::class, Ticket::class],
```
Comment (cconard96) [inline src/Glpi/Search/SearchEngine.php:220]:
```suggestion
```

This does not work as expected. "itil_types" are all CommonITILObject types so this adds ticket, change and problem as meta itemtypes for themselves while also not functioning with the "linked assistance objects". In fact meta Change on Ticket joins the "glpi_items_tickets" table where `glpi_items_tickets.itemtype = "Change"`.

File: tests/functional/Glpi/Search/SearchEngineTest.php [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+            'kanban_types',
+            'appliance_relation_types',
+            'remote_management_types',
+            'default_impact_asset_types',
+            'impact_asset_types',
+            'itemvirtualmachines_types',
+            'management_types',
+            'tools_types',
+            'admin_types',
+            'environment_types',
```
Comment (trasher) [inline tests/functional/Glpi/Search/SearchEngineTest.php:94]:
```suggestion
            'environment_types',
            'itil_types',
```

File: src/Glpi/Search/SearchEngine.php [hunk 6]
```diff
@@ -217,13 +217,16 @@ private static function getMetaParentItemtypesForTypesConfig(string $config_key)
...
+            'consumables_types'    => [\Consumable::class],
```
Comment (cconard96) [inline src/Glpi/Search/SearchEngine.php:221]:
```suggestion
```

Does not work as expected.
```sql
LEFT JOIN `glpi_consumables` ON (`glpi_users`.`id` = `glpi_consumables`.`items_id` AND `glpi_consumables`.`itemtype` = 'User' AND (`glpi_consumables`.`entities_id` IN ('0')))
```

File: src/Glpi/Search/SearchEngine.php [hunk 7]
```diff
@@ -217,13 +217,16 @@ private static function getMetaParentItemtypesForTypesConfig(string $config_key)
...
+            'process_types'        => [\Item_Process::class],
```
Comment (cconard96) [inline src/Glpi/Search/SearchEngine.php:222]:
```suggestion
```
This has no search options available for meta searches.

File: tests/functional/Glpi/Search/SearchEngineTest.php (hunk 3 above)
Comment (cconard96) [reply tests/functional/Glpi/Search/SearchEngineTest.php:47]: With so many exclusions, I wonder if a hard-coded list is better in addition to a new plugin hook so plugins can add new meta itemtypes.

File: tests/functional/Glpi/Search/SearchEngineTest.php [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+            'itil_types',
```
Comment (trasher) [inline tests/functional/Glpi/Search/SearchEngineTest.php]:
```suggestion
            'itil_types',
            'consumables_types',
            'process_types',
```

File: tests/functional/Glpi/Search/SearchEngineTest.php (hunk 3 above)
Comment (trasher) [reply tests/functional/Glpi/Search/SearchEngineTest.php:47]: I do not get the point. There are many exclusions because many of the "types" in config are not meta parent classes (as far as I know, this should not be used like that - but that is).
The test is just here to be sure any new "_type" entry is properly handled from that parent discovering point of view.

**Comment by trasher** [comment]
So, finally, what was not supported and now is (18 cases):
'state_types', 'document_types', 'reservation_types', 'location_types', 'link_types', 'networkport_types', 'contract_types', 'printer_types', 'certificate_types', 'cluster_types', 'operatingsystem_types', 'software_types', 'domain_types', 'databaseinstance_types', 'agent_types', 'line_types', 'itemantivirus_types', 'plug_types'.

**Comment by trasher** [comment]
ping?

**Review by cedric-anne — APPROVED** [review]
